### ISSUE TITLE
TypeError: mapped.substr is not a function (in validateMapping)

### ISSUE THREAD

**Issue by nlwillia** [systemjs/builder#529]
I encountered this error while trying out the [jspm-dev-builder](https://github.com/jackfranklin/jspm-dev-builder), but it looks like it may be a problem with the base builder or SystemJS.

This is with SystemJS 0.19.24.  jspm-dev-builder is currently setup by default against jspm 0.16, but for my 0.17 project I had updated the local jspm-dev-builder module's package.json to 0.17.  jspm-dev-builder doesn't appear to be doing anything terribly version-specific, and after I hacked around the error, it worked well enough, so I don't think this is related to the setup.

The program I was running is just a simple node process passing change events to the builder.
```
const DevBuilder = require('jspm-dev-builder');
const chokidar = require('chokidar');

var builder = new DevBuilder({
    expression: 'src/main',
    outLoc: 'dev-build.js',
    buildOptions: {
        sourceMaps: true,
        lowResSourceMaps: true
    }
});

builder.build();

chokidar.watch('**/*.ts').on('change', function(path) {
    builder.build(path);
});
```

The initial build succeeded, but as soon as I changed a file and the watcher triggered the builder, I got this error:
```
TypeError: mapped.substr is not a function
    at validateMapping (...\node_modules\jspm-dev-builder\node_modules\systemjs\dist\system.src.js:2141:31)
    at doMapSync (...\node_modules\jspm-dev-builder\node_modules\systemjs\dist\system.src.js:2148:5)
    at applyPackageConfigSync (...\node_modules\jspm-dev-builder\node_modules\systemjs\dist\system.src.js:2128:16)
    at SystemJSNodeLoader.<anonymous> (...\node_modules\jspm-dev-builder\node_modules\systemjs\dist\system.src.js:2339:14)
    at SystemJSNodeLoader.normalizeSync (...\node_modules\jspm-dev-builder\node_modules\systemjs\dist\system.src.js:4239:30)
    at SystemJSNodeLoader.loader.normalizeSync (...\node_modules\jspm-dev-builder\node_modules\systemjs-builder\lib\builder.js:223:42)
    at ...\node_modules\jspm-dev-builder\node_modules\systemjs-builder\lib\builder.js:92:29
    at Array.forEach (native)
    at Builder.invalidate (...\node_modules\jspm-dev-builder\node_modules\systemjs-builder\lib\builder.js:91:27)
    ...
```
All the dev-builder did was call invalidate on the file that changed.  That triggers a loop through the contents of the traceCache where each module name is normalized by the loader.  The error occurs during normalization when [validateMapping](https://github.com/systemjs/systemjs/blob/master/lib/package.js#L190) receives a value that is an object instead of a string.

It's not clear to me whether this is bad data due to a problem with a config file or loader, a bug in validation before calling `validateMapping` or something that `validateMapping` should be checking for itself.  There are a few places where it's called, and some of them check for string vs. object of the `mapped` parameter before calling while others don't.

The data involved at the point of error is:
- mapMatch = `"./process.js"`
- mapped = `{ "node": "./process-node.js" }` // definitely not a string!
- pkgName = `file:///.../jspm_packages/github/jspm/nodelibs-process@0.2.0-alpha`
- pkg = `{ "main": "./process.js",  "map": { "./process.js": { "node": "./process-node.js" } } }`
- loader = `SystemJSNodeLoader`

My project doesn't have any direct dependency on nodelibs-process, but I guess it's involved somehow in the builder environment, and everything in the cache gets checked on an invalidation.

I was able to work around the problem by changing `mapped.substr` to `(mapped.node || mapped).substr`, but that may not be the right solution.

(All ellipses added by me to shorten paths.)

**Comment by guybedford** [comment]
I've created a better error for this in https://github.com/systemjs/systemjs/commit/028293d5d8a729cdbec3b7ede6f340c56bc88971.

The actual bug itself should be resolved in https://github.com/systemjs/builder/commit/db663f0c69b5401eb215958ca0bdb758d673e223 now.

**Comment by guybedford** [comment]
Released in 0.15.14.
